FIX theming bugs

ndelangen opened the pull request:
Issue: https://github.com/storybooks/storybook/issues/5746

## What I did
- FIX the theme not showing on first load
- persist the theme into localstorage to prevent flash upon reload

I fixed this a bit quick and dirty, Will cleanup after V5 release

ndelangen committed:
FIX the themeing flash & persist the theme into localstorage to prevent flashing upon reloading

tmeasday reviewed (approved):
LGTM. Agreed that further refactoring into `core/ui` etc probably makes sense

ndelangen committed:
Merge branch 'next' into fix/theming-flash

# Conflicts:
#	examples/official-storybook/tests/__snapshots__/storyshots.test.js.snap

ndelangen committed:
REMOVE options from snapshot

ndelangen committed:
FIX problem with loading empty localStorage to theme

ndelangen committed:
CHANGE the maxDepth of stringify in postMessage channel as part of the theme was getting lost
ADD telejson as stringifier & parser for persistence
CLEANUP state initialization
CHANGE initial state for shortcuts so it's not dependent on this.getState()
DEDUPLICATE merge util function
FIX unnecessary messages about transition & brand missing in theme

ndelangen committed:
REMOVE dark theme

ndelangen committed:
FIX typescript issue

ndelangen commented:
@tmeasday Ended up having to introduce telejson in the persistence anyway, and had to up the maxDepth a bit for postMessage (not the entire theme was getting through)

tmeasday commented on lib/ui/src/core/context.js:
```diff
@@ -50,7 +50,6 @@ export class Provider extends Component {
     // Initialize the state to be the initial (persisted) state of the store.
     // This gives the modules the chance to read the persisted state, apply their defaults
     // and override if necessary
-    this.state = store.getInitialState();
```
I'm concerned this will break persistence. The idea (as expressed by the comment above) was that the store's persisted state gets passed into the modules.

ndelangen replied:
reverted

tmeasday commented on lib/ui/src/core/store.js:
```diff
@@ -29,11 +32,11 @@ export default class Store {
     // when you set keys, so it makes sense to do the same in combining the two storage modes
     // Really, you shouldn't store the same key in both places
-    return { ...get(localStorage), ...get(sessionStorage) };
+    return merge(base, { ...get(localStorage), ...get(sessionStorage) });
```
Again, I think this make break things. The idea was that modules would pick the relevant keys out of the persisted state and combine sensibly with their defaults, rather than us doing anything automatically via `merge`.

There are a bunch of ways this can break (doing it automatically).

ndelangen replied:
reverted

tmeasday reviewed (changes requested):
We decided not to persistence this way.. open to changing it in the future but not at this stage of the release.

tmeasday commented:
Ok with the rest of it, but if the issue was combining the theme defaults with the persisted theme, that should just be done the same way that e.g. versions does it.

ndelangen committed:
IMPROVE the default size of panels when switching bottom and right (minimal panelsize & maximum main size)

ndelangen committed:
MOVE setOptions to layout module
REMOVE merge from initial state
MOVE initial state of layout & ui to layout module
ADD merge of initial, restored to layout module

ndelangen commented on lib/ui/src/core/init-provider-api.js:
```diff
@@ -1,86 +1,6 @@
-import pick from 'lodash.pick';
```
This all moved to the 'layout' module as is.

ndelangen commented:
@tmeasday thanks for the comments, You're right, I changed it back.

ndelangen committed:
FIX theming bug

ndelangen committed:
FIX selectedPanel not being stored to persisted store

tmeasday commented on lib/ui/src/core/addons.js:
```diff
@@ -22,14 +22,18 @@ export default ({ provider, store }) => {
+      selectedPanel: ensurePanel(
+        api.getPanels(),
+        store.getState().selectedPanel,
+        store.getState().selectedPanel
```
This is a bit weerd (grabbing the selected panel twice). Is it intentional?

shilman replied:
Yeah, what's going on here?

ndelangen replied:
Just followed the arguments of `ensurePanel`,
I can remove 1 I think.

tmeasday commented on lib/ui/src/core/layout.js:
```diff
@@ -69,7 +130,54 @@ export default function({ store }) {
+            selectedPanel: options.panel || options.selectedPanel || selectedPanel,
+          },
+          { persistence: 'permanent' }
+        );
```
Does this mean we perma-persist all options, including the ones that were session-persisted above?

I'm OK with this for now for simplicity, but we might want to do something more subtle in the future (which will probably require improvements to the persistence API)

ndelangen replied:
Yeah I briefly experimented with persisting all the layout & ui options, but decided to not increase the size of this PR even more.

But it was actually really nice, things like full-screen, nav, panelPosition being preserved over refreshes...

On my todo list after V5 launch.

tmeasday commented on lib/ui/src/core/shortcuts.js:
```diff
@@ -31,7 +32,7 @@ export default function initShortcuts({ store }) {
     // Getting and setting shortcuts
     getShortcutKeys() {
-      return store.getState().shortcuts;
+      return store.getState().shortcuts || getInitialState().shortcuts;
```
@ndelangen do you remember why this line is necessary? I thought we put the initial/default state into the store right away on line 225 below?

ndelangen replied:
Might be unnecessary now, let me check..

tmeasday reviewed (approved):
I think this is OK but i have a couple of comments that we should probably discuss and resolve in future versions

tmeasday commented:
Thanks for pulling the layout module out @ndelangen !

ndelangen committed:
Merge branch 'next' into fix/theming-flash

ndelangen committed:
CLEANUP
